bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+                    rx = conn.rx;
+                }
+                ConnErr::Close => {
+                    // The Main task told us to shutdown
+                    return;
+                }
+            }
+        }
+    });
+    PeerConnHandle { handle, tx, addr, unique_id }
+}
+
+/// Spawn a task that handles accepted connections from a peer
+pub async fn spawn_server(
+    unique_id: u64,
+    my_peer_id: Baseboard,
+    my_addr: SocketAddrV6,
+    addr: SocketAddrV6,
+    sock: TcpStream,
+    main_tx: mpsc::Sender<ConnToMainMsg>,
+    log: &Logger,
+) -> AcceptedConnHandle {
+    // Create a channel for sending `MainToConnMsg`s to this connection task
+    let (tx, rx) = mpsc::channel(2);
+    let log = log.clone();
+    let handle = tokio::spawn(async move {
+        let (read_sock, write_sock, identify) = match perform_handshake(
+            sock,
+            &my_peer_id,
+            my_addr,
+            &log,
+        )
+        .await
+        {
+            Ok(val) => val,
+            Err(e) => {
+                warn!(log, "Handshake error: {:?}", e; "addr" => addr.to_string());
+                // This is a server so we bail and wait for a new connection
+                // We must inform the main task so it can clean up any metadata.
+                main_tx
+                    .send(ConnToMainMsg {
+                        handle_unique_id: unique_id,
+                        msg: ConnToMainMsgInner::FailedServerHandshake { addr },
+                    })
+                    .await;
+                return;
+            }
+        };
+
+        // Inform the main task that we have connected to a peer
+        let _ = main_tx
+            .send(ConnToMainMsg {
+                handle_unique_id: unique_id,
+                msg: ConnToMainMsgInner::ConnectedServer {
+                    accepted_addr: addr,
+                    addr: identify.addr.clone(),
+                    peer_id: identify.id.clone(),
+                },
+            })
+            .await;
+
+        let mut conn = EstablishedConn::new(
+            identify.id.clone(),
+            unique_id,
+            write_sock,
+            read_sock,
+            main_tx.clone(),
+            rx,
+            log.clone(),
+        );
+
+        // We always exit on server tasks, as the remote peer
+        // will reconnect.
+        let _ = conn.run().await;
+    });
+
+    AcceptedConnHandle { handle, tx, addr, unique_id }
+}
+
+// Serialize and write `msg` into `buf`, prefixed by a 4-byte big-endian size header
+//
+// Return the total amount of data written into `buf` including the 4-byte header
+fn write_framed<T: Serialize + ?Sized>(
+    msg: &T,
+) -> Result<Vec<u8>, ciborium::ser::Error<std::io::Error>> {
+    let mut cursor = Cursor::new(vec![]);
+    // Write a size placeholder
+    std::io::Write::write(&mut cursor, &[0u8; FRAME_HEADER_SIZE]);
+    cursor.set_position(FRAME_HEADER_SIZE as u64);
```

Comment:
Is this required? I think the cursor would be moved forward to `FRAME_HEADER_SIZE` by the `write` call above.

Reply:
I thought so as well, but that definitely was not the case.